### ISSUE TITLE
Benchmark positions for variants

### ISSUE THREAD

**Issue by ianfab** [ddugovic/Stockfish#158]
Since we have defined separate benchmark position sets for variants, I think we should collect some positions for each variant. In order to be useful for benchmarking, identifying (non-)functional patches, and debbuging, these positions should be diverse and cover topics like:
- opening
- middlegame
- early endgame
- late endgame
- tablebase positions (at least for atomic and giveaway)
- checkmate
- stalemate
- promotions
- en passant
- variant end (king in center in king of the hill, captured all pieces in giveaway, etc.)
- many pieces in hand in crazyhouse

Suggestions for positions and on topics to be added to/removed from the list are welcome.

**Comment by ddugovic** [comment]
No objections here!  I'll add the latest CVCA opening book...

**Comment by ianfab** [comment]
@ddugovic Do you mean the whole book or a sample? I think we should aim for something like 20-50 positions per variant (for standard chess there are 37), so something like 5-10 opening positions should suffice.

**Comment by ddugovic** [comment]
Oh, I had a different understanding of "benchmark positions" to include as many positions as possible.  (Sorry for my abbreviated comment, I need to run at the moment.)

**Comment by ianfab** [comment]
In principle it is of course desirable for a benchmark test to be as long as possible in order to get accurate results. However, since the duration of the benchmark test can easily be adjusted by increasing the depth, there is not really a need for a large number of positions (oh, we might have to adjust the default benchmark depth for variants, since antichess, crazyhouse, etc. can be very different in this respect). And sometimes you only want to do very short benchmark tests (e.g. to get the bench numbers), and then a large number of positions would be counterproductive.

As far as I can judge, it is only important that the positions are more or less representative of the positions that are typically searched so that the benchmark result is representative, and that also very rare events are triggered (like calling specialized endgame evaluation functions, even though not used for variants yet), so that bench numbers always change even if there is only a very minor functional change.

**Comment by ddugovic** [comment]
Excellent point, well explained, and I agree.  Sorry I failed to hint my support for this idea in my previous comment.

**Comment by ddugovic** [comment]
@ianfab is it OK if I assign you on this issue?

**Comment by ianfab** [comment]
@ddugovic Yes, sure.

**Comment by QueensGambit** [comment]
Hello Fabian and Daniel,

I agree that benchmark positions are very helpful because they can be used to evaluate patches much quicker and be used for some sort of unit test.

https://github.com/noelben also suggested creating a benchmark database for positions earlier.
I've already collected some positions manually for crazyhouse with different motives.

Multi-variant-Stockfish and CrazyAra seem to have very different weaknesses at the moment:

Given an example position taken from this game: https://lichess.org/xGrMuwnm#16
```
rnbqk2r/ppp2ppp/4p3/3N4/3P4/5N2/PpPQ1PPP/R3KB1R/BNPpb w KQkq - 0 9
```

It takes Stockfish 9/10 depth >19 to find the move Rd1 instead of Nxc7, but only 1 node for CrazyAra because it's a very human like move.
CrazyAra on the other hand has currently the problem of resolving certain tactics which are hard to resolve for humans as well.
I think it would make sense to give ground truth evaluations for a position using a diverse set of engines.
One major challenge will be creating a good standardized template on how to evaluate a benchmark position and how people can contribute a new position.
In certain positions where a mate in X is possible, there's clearly one or multiple best moves, but on more strategic positions (especially the opening) it's hard to tell what's really the best move in the long term. 
Also if an engine proposes a novelty in a test position it might not always be a bad move.
It could be useful to define blunder moves instead of good moves in certain positions.

Would it make sense to collaboratively create such a database in a new repository?

I've found different groups who have already worked on s.th. similar likes this for crazyhouse.
* https://chessvariants.training/Puzzle/Crazyhouse
* https://lichess.org/forum/team-crazyhouse-tactics-book-project

Despite the usefulness of such a benchmark positions test suit, conducting real test matches still remains a necessity because there's the danger that the engine will implicitly overfit to the test position ground truth targets.

**Comment by ianfab** [comment]
@QueensGambit The "benchmark" positions in SF are not really for benchmarking playing strength or evaluation accuracy, they are just supposed to be a diverse set of positions which can be used to measure average speed (nps) and to identify whether a patch introduces a functional change or not.

Having dedicated test suites to evaluate strength and weaknesses might also be useful, but from my experience they can only help to identify problematic positions that could help to give you some ideas for improvements (when analyzing them one by one manually), whereas they are not very useful for measuring playing strength. So although such a collection sounds like a nice idea, I do not really see how to directly make use of it it practice.

**Comment by QueensGambit** [comment]
@ianfab Thanks for making this clear. Your arguments are reasonable.

**Comment by bftjoe** [comment]
bench for variants finishes quite quickly on modern CPU. Might want to either increase depth or add more positions to make them comparable to more easily measure performance.

Stockfish-mv built for atomic only on i9 9900k

regular chess

47 positions

Total time (ms) : 2555
Nodes searched  : 4074430
Nodes/second    : 1594688

Atomic bench

10 positions

Total time (ms) : 854
Nodes searched  : 1839492
Nodes/second    : 2153971

**Comment by ddugovic** [comment]
FYI for performance reasons (timeouts) I had to remove "puzzle" tests from the Travis CI build (thankfully, these tests can still run on GitHub Actions):
https://travis-ci.org/github/ddugovic/Stockfish/branches

I think there are many good reasons to have benchmarks, and to add more positions... ironically these days I haven't been using bench so much to measure performance, but instead to verify builds are stable.
